GameServerLib/GameObjects/AttackableUnits/AI/ObjAIBase.cs

```diff
@@ -632,6 +552,147 @@ public bool RecalculateAttackPosition()
+                SetWaypoints(new List<Vector2>() { GetPosition(), TargetUnit.GetPosition() });
+                /* TODO: Soon we will use path finding for this.
+                 * if(CurWaypoint >= Waypoints.Count)
+                {
+                    var newWaypoints = _game.Map.NavGrid.GetPath(GetPosition(), TargetUnit.GetPosition());
+                    if (newWaypoints.Count > 1)
+                    {
+                        SetWaypoints(newWaypoints);
+                    }
+                }*/
+            }
+        }
+
+        public void RemoveBuff(IBuff b)
+        {
+            lock (_buffsLock)
+            {
+                if (b.BuffAddType == BuffAddType.STACKS_AND_OVERLAPS && b.StackCount > 1)
+                {
+                    b.DecrementStackCount();
+
+                    RemoveBuff(b.Name);
+                    BuffList.Remove(b);
+                    RemoveBuffSlot(b);
+
+                    var tempbuffs = GetBuffsWithName(b.Name);
+
+                    for (int i = 0; i < tempbuffs.Count; i++)
+                    {
+                        tempbuffs[i].SetStacks(b.StackCount);
+                    }
+
+                    BuffSlots[b.Slot] = tempbuffs[0];
+                    Buffs.Add(b.Name, tempbuffs[0]);
+
+                    var newestBuff = tempbuffs[tempbuffs.Count - 1];
+
+                    if (!b.IsHidden)
+                    {
+                        if (b.BuffType == BuffType.COUNTER)
+                        {
+                            _game.PacketNotifier.NotifyNPC_BuffUpdateNumCounter(Buffs[b.Name]);
+                        }
+                        else
+                        {
+                            if (b.StackCount == 1)
+                            {
+                                _game.PacketNotifier.NotifyNPC_BuffUpdateCount(newestBuff, b.Duration - newestBuff.TimeElapsed, newestBuff.TimeElapsed);
+                            }
+                            else
+                            {
+                                _game.PacketNotifier.NotifyNPC_BuffUpdateCountGroup(this, tempbuffs, b.Duration - newestBuff.TimeElapsed, newestBuff.TimeElapsed);
+                            }
+                        }
+                    }
+                }
+                else
+                {
+                    BuffList.RemoveAll(buff => buff.Elapsed());
+                    RemoveBuff(b.Name);
+                    RemoveBuffSlot(b);
+                    if (!b.IsHidden)
+                    {
+                        _game.PacketNotifier.NotifyNPC_BuffRemove2(b);
+                    }
+                }
+            }
+        }
+
+        public void RemoveBuff(string b)
+        {
+            lock (_buffsLock)
+            {
+                Buffs.Remove(b);
+            }
+        }
+
+        public void RemoveBuffsWithName(string buffName)
+        {
+            lock (_buffsLock)
+            {
+                for (int i = 0; i < BuffList.Count; i++)
```

Comment:
Use LINQ statement like `BuffList.Where(x => x.IsBuffSame(buffname)).ForEach(x => x.DeactivateBuff())` instead of for.